src/psyclone/psyir/transformations/omp_loop_trans.py

```diff
@@ -0,0 +1,261 @@
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: S. Siso, STFC Daresbury Lab
+
+''' Transformation to insert OpenMP directives to parallelise PSyIR Loops. '''
+
+from psyclone.configuration import Config
+from psyclone.psyir.nodes import Routine, OMPDoDirective, OMPLoopDirective, \
+    OMPParallelDoDirective, OMPTeamsDistributeParallelDoDirective
+from psyclone.psyir.symbols import DataSymbol, INTEGER_TYPE
+from psyclone.psyir.transformations.parallel_loop_trans import \
+    ParallelLoopTrans
+
+VALID_OMP_SCHEDULES = ["runtime", "static", "dynamic", "guided", "auto"]
+MAP_STR_TO_LOOP_DIRECTIVES = {
+    "do": OMPDoDirective,
+    "paralleldo": OMPParallelDoDirective,
+    "teamsdistributeparalleldo": OMPTeamsDistributeParallelDoDirective,
+    "loop": OMPLoopDirective
+}
+VALID_OMP_DIRECTIVES = list(MAP_STR_TO_LOOP_DIRECTIVES.keys())
+
+
+class OMPLoopTrans(ParallelLoopTrans):
+    '''
+    Adds an OpenMP directive to parallelise this loop. It can insert different
+    directives such as "omp do/for", "omp parallel do/for", "omp teams
+    distribute parallel do/for" or "omp loop" depending on the provided
+    parameters.
+    The OpenMP schedule to use can also be specified, but this will be ignored
+    in case of the "omp loop". The configuration-defined 'reprod' parameter
+    also specifies whether a manual reproducible reproduction is to be used.
+    Note, reproducible in this case means obtaining the same results with the
+    same number of OpenMP threads, not for different numbers of OpenMP threads.
+
+    :param str omp_schedule: the OpenMP schedule to use. Defaults to 'auto'.
+    :param str omp_directive: choose which OpenMP loop directive to use. \
+        Defaults to "omp do"
+
+    For example:
+
+    >>> from psyclone.psyir.frontend.fortran import FortranReader
+    >>> from psyclone.psyir.backend.fortran import FortranWriter
+    >>> from psyclone.psyir.nodes import Loop
+    >>> from psyclone.transformations import OMPLoopTrans, OMPParallelTrans
+    >>>
+    >>> psyir = FortranReader().psyir_from_source("""
+    ...     subroutine my_subroutine()
+    ...         integer, dimension(10, 10) :: A
+    ...         integer :: i
+    ...         integer :: j
+    ...         do i = 1, 10
+    ...             do j = 1, 10
+    ...                 A(i, j) = 0
+    ...             end do
+    ...         end do
+    ...     end subroutine
+    ...     """)
+    >>> loop = psyir.walk(Loop)[0]
+    >>> omplooptrans1 = OMPLoopTrans(omp_schedule="dynamic",
+    ...                              omp_directive="paralleldo")
+    >>> omplooptrans1.apply(loop)
+    >>> print(FortranWriter()(psyir))
+    subroutine my_subroutine()
+      integer, dimension(10,10) :: a
+      integer :: i
+      integer :: j
+      integer :: th_idx
```

Comment:
doctest fails for this docstring because `th_idx` and `nthreads` aren't in the generated code.

Reply:
Removed variables. Now the doctest should pass